libtenzir/builtins/operators/where_map.cpp

```diff
@@ -677,8 +682,87 @@ class assert_plugin final : public virtual operator_factory_plugin {
   }
 };
 
+// TODO: Don't want to write this fully ourselves.
+class where_exec final : public exec::operator_base {
+public:
+  where_exec() = default;
+
+  auto name() const -> std::string override {
+    return "where_exec";
+  }
+
+  explicit where_exec(ast::expression predicate)
+    : predicate_{std::move(predicate)} {
+  }
+
+  friend auto inspect(auto& f, where_exec& x) -> bool {
+    return f.apply(x.predicate_);
+  }
+
+private:
+  ast::expression predicate_;
+};
+
+// TODO: Don't want to write this fully ourselves.
+class where_ir final : public ir::operator_base {
+public:
+  where_ir() = default;
+
+  where_ir(location self, ast::expression predicate)
+    : self_{self}, predicate_{std::move(predicate)} {
+  }
+
+  auto name() const -> std::string override {
+    return "where_ir";
+  }
+
+  auto substitute(substitute_ctx ctx, bool instantiate)
+    -> failure_or<void> override {
+    (void)instantiate;
+    TRY(predicate_.substitute(ctx));
+    return {};
+  }
+
+  // TODO: Should this get the type of the input?
+  // Or do we get it earlier? Or later?
+  auto finalize(finalize_ctx ctx) && -> failure_or<exec::pipeline> override {
+    (void)ctx;
+    return std::make_unique<where_exec>(std::move(predicate_));
+  }
+
+  auto infer_type(operator_type2 input, diagnostic_handler& dh) const
+    -> failure_or<std::optional<operator_type2>> override {
+    if (input.is_not<table_slice>()) {
+      // TODO: Do not duplicate these messages across the codebase.
+      diagnostic::error("operator expects events").primary(self_).emit(dh);
+      return failure::promise();
+    }
+    return tag_v<table_slice>;
+  }
```

Comment:
To avoid this from happening, did you consider returning a list of pairs indicating the mapping of input to output types? Something like this:

```cpp
// maps table slices to table slices, but nothing else
return {
  {tag_v<table_slice>, tag_v<table_slice>},
};
```

```cpp
// maps both chunks and table slices to table slices
return {
  {tag_v<chunk_ptr>, tag_v<table_slice>},
  {tag_v<table_slice>, tag_v<table_slice>},
};
```

Reply:
I had something like that initially, but then noticed that you sometimes want to report errors deeper in the chain. Like for example in nested pipelines. And you might have custom error messages there. For example, `if` requires that the output type of the two branches match. Thus I pulled the error messages into the implementation. We can refactor this (into something like `make_type_clash_error()` or whatever). Just didn't think that this was necessary yet, but I would definitely do this at a later point in time if the error reporting stays in the implementation.